### PR TITLE
SI-7785 Preserve TypeVar suspension through TypeMaps

### PR DESCRIPTION
**Pull request by retronym**
During `findMember`, TypeVars in `this` are placed into suspended
animation. This is to avoid running into recursive types when
matching members to those in base classes.

However, the mechanism used to do this is superficial, and doesn't
work when TypeVars are copied by TypeMaps. This seems to crop up
when using `AppliedTypeVar` with higher-kinded type vars.

In the enclosed test case, the cyclic type led to a SOE in
CommonOwnerMap.

This commit allows a TypeVar to delegate its `suspended` attribute
to the TypeVar from which it was copied. This is done in
`TypeVar#applyArgs`, which is called by:

```
// TypeMap#mapOver
case tv@TypeVar(_, constr) =>
  if (constr.instValid) this(constr.inst)
  else tv.applyArgs(mapOverArgs(tv.typeArgs, tv.params))
```

We should review the other places this is called to make sure
that it make sense to link in this way:

```
Types#appliedType
TypeVar#normalize
```

Review by @adriaanm.

Even though this was a regression in 2.10.1, I'm submitting this to 2.11 as the change isn't without risk, and the only reason this worked in 2.10.0 was because of another bug (`TypeVar` had a structural equals.)

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -3065,7 +3069,16 @@ trait Types
     // </region>
 
     // ignore subtyping&equality checks while true -- see findMember
-    private[Types] var suspended = false
+    // OPT: This could be Either[TypeVar, Boolean], but this encoding was chosen instead to save allocations.
+    private var _suspended: Type = TypeVar.ConstantFalse
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:3073]: How about introducing two sentinel `TypeVar`s, and narrowing `_suspended`'s type to `TypeVar`?

Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala:3073]: Yeah, I guess that's a fraction less ugly than this :)

**Comment by Blaisorblade** [comment]
I'd like that backported to 2.10 (BTW, I observed this as a regression in 2.10.2, 2.10.1 was fine), and it's a compiler crash. But if the risk/benefit ratio is as you describe, I'm fine with not having a backport (at least if I can workaround all other instances, I'll take a look).

**Comment by Blaisorblade** [comment]
> We should review the other places this is called [...]

OK, I'd ask for a backport if and only if that review gets done.

**Comment by retronym** [comment]
It worked in 2.10.1 by a total fluke, as different instances of `TypeVar`-s were considered equal so some `TypeMap`-s where considred to be identities and returned the originals. I've chased down one-too-many regressions in 2.10.x, often from code that itself was fine but unmasked existing bugs, to be confident to backport this one.

**Comment by adriaanm** [comment]
As a general policy, after a couple of minor releases, we focus on the next major version. At this point, we won't do back ports (ourselves) except for critical issues and where the risk of regression is low. Contributions welcome, as always, though the regression clause still holds of course.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala:3073]: Actually that created an ugly cycle: the constructor of the sentinel TypeVar tries to assign the sentinel to `_suspended`, which constucts a new sentinel, which...

```
    at scala.reflect.internal.Types$TypeVar$.sentinelTypeVar(Types.scala:2851)
    at scala.reflect.internal.Types$TypeVar$.<init>(Types.scala:2852)
    at scala.reflect.internal.SymbolTable.TypeVar$lzycompute(SymbolTable.scala:14)
    at scala.reflect.internal.SymbolTable.TypeVar(SymbolTable.scala:14)
    at scala.reflect.internal.Types$TypeVar$.sentinelTypeVar(Types.scala:2851)
    at scala.reflect.internal.Types$TypeVar$.<init>(Types.scala:2852)
    at scala.reflect.internal.SymbolTable.TypeVar$lzycompute(SymbolTable.scala:14)
    at scala.reflect.internal.SymbolTable.TypeVar(SymbolTable.scala:14)
```

I'm inclined to leave it as is.

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:3073]: Ok, thanks for trying it out!

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**